# Logistic Regression

## Dataset 
* Wolberg,WIlliam. (1992). Breast Cancer Wisconsin (Original). UCI Machine Learning Repository. https://doi.org/10.24432/C5HP4Z.
* Dataset has been preprocessed to remove the null values

## Importing the libraries

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

## Importing the dataset

In [2]:
dataset = pd.read_csv("breast_cancer.csv")

In [3]:
dataset.head()

Sample code number  Clump Thickness  Uniformity of Cell Size  \
0             1000025                5                        1   
1             1002945                5                        4   
2             1015425                3                        1   
3             1016277                6                        8   
4             1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

   Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0            1                3                1        1      2  
1           10                3                2        1      2  
2            2                3                1        1      2  
3            4                3                7        1      2  
4            1                3                1        1      2

In [4]:
dataset.isnull().sum()

Sample code number             0
Clump Thickness                0
Uniformity of Cell Size        0
Uniformity of Cell Shape       0
Marginal Adhesion              0
Single Epithelial Cell Size    0
Bare Nuclei                    0
Bland Chromatin                0
Normal Nucleoli                0
Mitoses                        0
Class                          0
dtype: int64

In [5]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 683 entries, 0 to 682
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   Sample code number           683 non-null    int64
 1   Clump Thickness              683 non-null    int64
 2   Uniformity of Cell Size      683 non-null    int64
 3   Uniformity of Cell Shape     683 non-null    int64
 4   Marginal Adhesion            683 non-null    int64
 5   Single Epithelial Cell Size  683 non-null    int64
 6   Bare Nuclei                  683 non-null    int64
 7   Bland Chromatin              683 non-null    int64
 8   Normal Nucleoli              683 non-null    int64
 9   Mitoses                      683 non-null    int64
 10  Class                        683 non-null    int64
dtypes: int64(11)
memory usage: 58.8 KB


In [6]:
dataset.describe()

Sample code number  Clump Thickness  Uniformity of Cell Size  \
count        6.830000e+02       683.000000               683.000000   
mean         1.076720e+06         4.442167                 3.150805   
std          6.206440e+05         2.820761                 3.065145   
min          6.337500e+04         1.000000                 1.000000   
25%          8.776170e+05         2.000000                 1.000000   
50%          1.171795e+06         4.000000                 1.000000   
75%          1.238705e+06         6.000000                 5.000000   
max          1.345435e+07        10.000000                10.000000   

       Uniformity of Cell Shape  Marginal Adhesion  \
count                683.000000         683.000000   
mean                   3.215227           2.830161   
std                    2.988581           2.864562   
min                    1.000000           1.000000   
25%                    1.000000           1.000000   
50%                    1.000000           1.000000   
75%                    5.000000           4.000000   
max                   10.000000          10.000000   

       Single Epithelial Cell Size  Bare Nuclei  Bland Chromatin  \
count                   683.000000   683.000000       683.000000   
mean                      3.234261     3.544656         3.445095   
std                       2.223085     3.643857         2.449697   
min                       1.000000     1.000000         1.000000   
25%                       2.000000     1.000000         2.000000   
50%                       2.000000     1.000000         3.000000   
75%                       4.000000     6.000000         5.000000   
max                      10.000000    10.000000        10.000000   

       Normal Nucleoli     Mitoses       Class  
count       683.000000  683.000000  683.000000  
mean          2.869693    1.603221    2.699854  
std           3.052666    1.732674    0.954592  
min           1.000000    1.000000    2.000000  
25%           1.000000    1.000000    2.000000  
50%           1.000000    1.000000    2.000000  
75%           4.000000    1.000000    4.000000  
max          10.000000   10.000000    4.000000

## Splitting the dataset into Training set and Test set

In [7]:
X = dataset.iloc[:,1:-1].values
Y = dataset.iloc[:,-1].values

In [8]:
# from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

* X_train = matrix of features conatining the observations of the training set
* X_test = matrix of features conatining the observations of the test set
* Y_train = dependent variable vector conatining the observations of the tarining set
* Y_test = dependent variable vector conatining the observations of the test set

## Training the Logistic Regression model on the training set

In [9]:
# from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, Y_train)

LogisticRegression(random_state=0)

## Predicting the test set results

In [10]:
Y_pred = classifier.predict(X_test)

## Making the Confusion Matrix

In [11]:
# from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)
print(cm)

[[84  3]
 [ 3 47]]


* We have 84 correct predictions that the tumor is Benign (True Positives)
* We have 47 correct predictions that the tumor is Malignant (True Negatives)
* We have 3 incorrect predictions that the tumor is Benign (False Positives)
* We have 3 incorrect predictions that the tumor is Malignant (False Negatives)

## Computing the accuracy with k-Fold Cross Validation

In [12]:
# from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = Y_train, cv = 10)
print("Accuracies: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100)) # Standard Deviation of the 10 accuracies

Accuracies: 96.70 %
Standard Deviation: 1.97 %


* Out of 100 observations almost 97 will be correct
* Standard Deviation of 1.97% means the ten accuracies more or less will fall between 96.7-1.97 % and 96.7+1.97 %